src/chainstate/stacks/index/marf.rs

```diff
@@ -525,40 +554,44 @@ impl MARF {
     pub fn insert_batch(&mut self, keys: &Vec<String>, values: Vec<MARFValue>) -> Result<(), Error> {
         assert_eq!(keys.len(), values.len());
 
-        match self.open_chain_tip {
+        let block_hash = match self.open_chain_tip {
             None => {
                 Err(Error::WriteNotBegunError)
             },
-            Some(ref block_hash) => {
-                if keys.len() == 0 {
-                    return Ok(());
-                }
+            Some(WriteChainTip{ ref block_hash, .. }) => {
+                Ok(block_hash.clone())
+            }
+        }?;
+
+        if keys.len() == 0 {
+            return Ok(());
+        }
         
-                let cur_block_hash = self.storage.get_cur_block();
-                let cur_block_rw = self.storage.readwrite();
-                
-                let mut i = 0;
-                while i < keys.len() - 1 {
-                    let key = &keys[i];
-                    let value = &values[i];
+        let cur_block_hash = self.storage.get_cur_block();
+        let cur_block_rw = self.storage.readwrite();
                 
-                    let marf_leaf = TrieLeaf::from_value(&vec![], value.clone());
-                    let path = TriePath::from_key(key);
-
-                    MARF::insert_leaf_in_batch(&mut self.storage, block_hash, &path, &marf_leaf)?;
-                    i += 1;
-                }
+        let last = keys.len() - 1;
 
-                // last insert updates the root with the skiplist hash
-                let marf_leaf = TrieLeaf::from_value(&vec![], values[i].clone());
-                let path = TriePath::from_key(&keys[i]);
-                MARF::insert_leaf(&mut self.storage, block_hash, &path, &marf_leaf)?;
+        
+        let mut result = keys[0..last].iter().zip(values[0..last].iter())
+            .try_for_each(|(key, value)| {
+                let marf_leaf = TrieLeaf::from_value(&vec![], value.clone());
+                let path = TriePath::from_key(key);
+                        
+                MARF::insert_leaf_in_batch(&mut self.storage, &block_hash, &path, &marf_leaf)
+            });
+
+        if result.is_ok() {
+            // last insert updates the root with the skiplist hash
+            let marf_leaf = TrieLeaf::from_value(&vec![], values[last].clone());
+            let path = TriePath::from_key(&keys[last]);
+            result = MARF::insert_leaf(&mut self.storage, &block_hash, &path, &marf_leaf);
+        }
 
                 // restore
-                self.storage.open_block(&cur_block_hash, cur_block_rw)?;
-                Ok(())
-            }
-        }
+        self.storage.open_block(&cur_block_hash, cur_block_rw)?;
```

Comment:
Because the MARF insertion/read operations are supposed to provide a guarantee to their caller that the same block will be open at the end of the operation as at the beginning of the operation. Because there are paths where a "not found" read results in a `index::Error::NotFoundError`, the previous iteration of this would leave the MARF in a bad state following a not found read.